finagle-http/src/main/scala/com/twitter/finagle/http/HeaderMap.scala

```diff
@@ -52,67 +54,88 @@ abstract class HeaderMap
   def += (kv: (String, Date)): HeaderMap =
     += ((kv._1, HeaderMap.format(kv._2)))
 
-  override def empty: HeaderMap = new MapHeaderMap(mutable.Map.empty)
+  override def empty: HeaderMap = MapHeaderMap()
 }
 
+private[finagle] case class HeaderValuePair(header: String, value: String) {
+  val canonicalName = HeaderValuePair.canonicalName(header)
+}
+
+private[finagle] object HeaderValuePair {
+  def canonicalName(header: String) = header.toLowerCase
+}
 
 /** Mutable-Map-backed [[HeaderMap]] */
-class MapHeaderMap(underlying: mutable.Map[String, Seq[String]]) extends HeaderMap {
+class MapHeaderMap extends HeaderMap {
+
+  private val underlying = mutable.Map.empty[String, List[HeaderValuePair]]
+
+  // This is here to provide backward compatibility
+  def this(underlying: mutable.Map[String, Seq[String]]) {
```

Comment:
hmm i realize why you left it, but i think its a bit deceptive to the end-user. given that they pass in a mutable map, there is likely some expectation that their modifications to the map are reflected in this `HeaderMap`. but they won't.

this is a big enough change in behavior that i think we should at a minimum mark this as deprecated, and strongly consider breaking the API by removing the constructor.